File: freight-triage-polyglot-ledger-recovery/environment/native/src/format/ordinal_suffix.cpp
#include "freight/formats.h"

#include <algorithm>
#include <cstdio>

namespace freight {

// Formatter: ordinal suffix.
std::string fmt_ordinal_suffix(long long value) {
  long long mod100 = ((value % 100) + 100) % 100;
  long long mod10 = mod100 % 10;
  const char* suffix = "th";
  if (mod100 < 11 || mod100 > 13) {
    if (mod10 == 1) {
      suffix = "st";
    } else if (mod10 == 2) {
      suffix = "nd";
    } else if (mod10 == 3) {
      suffix = "rd";
    }
  }
  return std::to_string(value) + suffix;
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/format/OrdinalSuffix.java
package com.freight.format;

/** Formatter: ordinal suffix. */
public final class OrdinalSuffix implements Formatter {

    private static final String BASE36_DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    @Override
    public String name() {
        return "ordinal_suffix";
    }

    @Override
    public String apply(long value) {
        long mod100 = ((value % 100L) + 100L) % 100L;
        long mod10 = mod100 % 10L;
        String suffix = "th";
        if (mod100 < 11L || mod100 > 13L) {
            if (mod10 == 1L) {
                suffix = "st";
            } else if (mod10 == 2L) {
                suffix = "nd";
            } else if (mod10 == 3L) {
                suffix = "rd";
            }
        }
        return Long.toString(value) + suffix;
    }
}
